File: libs/ai-kernel/src/domain/rag/chunker.rs
pub fn chunk_text(content: &str, max_chars: usize) -> Vec<(i32, String)> {
    let mut chunks = Vec::new();
    let mut buffer = String::new();
    let mut position = 0;

    for paragraph in content.split("\n\n") {
        let trimmed = paragraph.trim();
        if trimmed.is_empty() {
            continue;
        }

        if buffer.len() + trimmed.len() + 2 > max_chars && !buffer.is_empty() {
            chunks.push((position, buffer.trim().to_string()));
            position += 1;
            buffer.clear();
        }

        if trimmed.len() > max_chars {
            for sentence in trimmed.split('.') {
                let sentence = sentence.trim();
                if sentence.is_empty() {
                    continue;
                }

                if buffer.len() + sentence.len() + 2 > max_chars && !buffer.is_empty() {
                    chunks.push((position, buffer.trim().to_string()));
                    position += 1;
                    buffer.clear();
                }

                buffer.push_str(sentence);
                buffer.push_str(". ");
            }
        } else {
            buffer.push_str(trimmed);
            buffer.push_str("\n\n");
        }
    }

    if !buffer.trim().is_empty() {
        chunks.push((position, buffer.trim().to_string()));
    }

    chunks
}

File: openfoundry-go/libs/ai-kernel-go/domain/rag/chunker.go
package rag

import "strings"

// Chunk is the (position, text) pair returned by ChunkText.
type Chunk struct {
	Position int32
	Text     string
}

// ChunkText splits a document into chunks no larger than maxChars.
// Algorithm mirrors Rust src/domain/rag/chunker.rs:
//   - Split on "\n\n" paragraphs.
//   - When the next paragraph would push the buffer past maxChars
//     (and the buffer is non-empty), flush the buffer as a chunk
//     and start fresh.
//   - When a single paragraph itself exceeds maxChars, split on
//     "." sentences and apply the same buffer-flush rule.
//   - Trim each emitted chunk and assign sequential positions
//     starting at 0.
func ChunkText(content string, maxChars int) []Chunk {
	chunks := []Chunk{}
	var buffer strings.Builder
	position := int32(0)

	for _, paragraph := range strings.Split(content, "\n\n") {
		trimmed := strings.TrimSpace(paragraph)
		if trimmed == "" {
			continue
		}

		if buffer.Len()+len(trimmed)+2 > maxChars && buffer.Len() > 0 {
			chunks = append(chunks, Chunk{Position: position, Text: strings.TrimSpace(buffer.String())})
			position++
			buffer.Reset()
		}

		if len(trimmed) > maxChars {
			for _, sentence := range strings.Split(trimmed, ".") {
				sentence = strings.TrimSpace(sentence)
				if sentence == "" {
					continue
				}
				if buffer.Len()+len(sentence)+2 > maxChars && buffer.Len() > 0 {
					chunks = append(chunks, Chunk{Position: position, Text: strings.TrimSpace(buffer.String())})
					position++
					buffer.Reset()
				}
				buffer.WriteString(sentence)
				buffer.WriteString(". ")
			}
		} else {
			buffer.WriteString(trimmed)
			buffer.WriteString("\n\n")
		}
	}

	if strings.TrimSpace(buffer.String()) != "" {
		chunks = append(chunks, Chunk{Position: position, Text: strings.TrimSpace(buffer.String())})
	}
	return chunks
}
